Stop returning nullptr for system font keywords on non-font related properties

chirags27 opened the pull request:
#### db5cab0a4aaaf397271d4422ae417a6f39ea18cb
<pre>
Stop returning nullptr for system font keywords on non-font related properties

<a href="https://bugs.webkit.org/show_bug.cgi?id=250245">https://bugs.webkit.org/show_bug.cgi?id=250245</a>
rdar://103916548

Reviewed by Darin Adler and Tim Nguyen.

The code that is handling font: menu shouldn’t apply to non-font
properties.

* LayoutTests/fast/css/css-text-border-menu-expected.txt: Added.
* LayoutTests/fast/css/css-text-border-menu.html: Added.
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::getPropertyCSSValue const):

Canonical link: <a href="https://commits.webkit.org/258742@main">https://commits.webkit.org/258742@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dd6f5b3d5e15a7f65e81e12d41adb3067e14dff9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102683 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11806 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35712 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111945 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172188 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106650 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12815 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2712 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94946 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109638 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108459 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9830 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37486 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/91682 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24554 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79229 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5263 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25973 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5417 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2423 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11438 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45462 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6007 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7154 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -952,6 +952,8 @@ String StyleProperties::getCommonValue(const StylePropertyShorthand& shorthand)
     for (auto longhand : shorthand) {
         auto value = getPropertyCSSValue(longhand);
+        if (!value)
+            return String();
```
If the bug is that https://searchfox.org/wubkat/rev/bfe2441d9e38a24d34f9bf9a8de6a19b07dd20a6/Source/WebCore/css/StyleProperties.cpp#1140-1143 is returning nullptr, even for non font-related properties (e.g. border), then the fix should be to check that the corresponding shorthand for the property in that codepath is font.

E.g.:

`shorthandForProperty(property.id()) == fontShorthand()`

I think this is an actual bug, `border: menu` successfully parses because `menu` is a color keyword, but at the same time, `menu` is a system font keyword, so the codepath above accidentally matches.

nt1m replied:
Sorry `shorthandForProperty` is the wrong method to use.

I think changing the code linked in the previous comment to:
```
    // System fonts are represented as CSSPrimitiveValue for various font subproperties, but these must serialize as the empty string.
    // It might be better to implement this as a special CSSValue type instead of turning them into null here.
    if (property.shorthandID() == CSSPropertyFont && CSSPropertyParserHelpers::isSystemFontShorthand(valueID(value)))
        return nullptr;
```
should work?

Anyway the idea is that the check should only return true for font longhand properties, not any other, like `border`

darinadler replied:
I agree that the change from
@nt1m looks like the correct fix.

chirags27 replied:
Sounds good, thanks for the feedback, I'll update my change to address that.

nt1m commented on LayoutTests/fast/css/css-text-unknown-element.html:
```diff
@@ -0,0 +1,13 @@
+</script>
+<body onload=jsfuzzer()>
+<p>PASS if no crash</p>
+</body>
```
```
<script>
function test() {
    if (window.testRunner)
        testRunner.dumpAsText();
    const div = document.createElement("div");
    const elementStyle = div.style;
    elementStyle.border = "menu";
    cssText = elementStyle.cssText;
}
</script>
<body onload="test()">
<p>PASS if no crash</p>
```

nit: can you use 4 space indentation, and format the test properly? The snippet above should repro the crash.

nt1m reviewed (changes requested):
I don't think this is the right fix.

nt1m commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -1063,7 +1064,9 @@ String StyleProperties::borderPropertyValue(const StylePropertyShorthand& width,
-                if (!getPropertyCSSValue(longhand)->isImplicitInitialValue())
+                auto propertyCSSValue = getPropertyCSSValue(longhand);
+                ASSERT(propertyCSSValue);
+                if (!propertyCSSValue->isImplicitInitialValue())
```
```suggestion
                auto value = getPropertyCSSValue(longhand);
                ASSERT(value);
                if (!value->isImplicitInitialValue())
```

nt1m reviewed (approved):
r=me if EWS is green.

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -1063,7 +1064,9 @@ String StyleProperties::borderPropertyValue(const StylePropertyShorthand& width,
             for (auto longhand : shorthand) {
-                if (!getPropertyCSSValue(longhand)->isImplicitInitialValue())
+                auto value = getPropertyCSSValue(longhand);
+                ASSERT(value);
```
Please don't add this.

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -952,6 +952,7 @@ String StyleProperties::getCommonValue(const StylePropertyShorthand& shorthand)
     String result;
     for (auto longhand : shorthand) {
         auto value = getPropertyCSSValue(longhand);
+        ASSERT(value);
```
Please don't add this.

darinadler reviewed (approved):
Looks good assuming the tests all still pass.

chirags27 committed:
Stop returning nullptr for system font keywords on non-font related properties

https://bugs.webkit.org/show_bug.cgi?id=250245
rdar://103916548

Reviewed by Darin Adler and Tim Nguyen.

The code that is handling font: menu shouldn’t apply to non-font
properties.

* LayoutTests/fast/css/css-text-border-menu-expected.txt: Added.
* LayoutTests/fast/css/css-text-border-menu.html: Added.
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::getPropertyCSSValue const):

Canonical link: https://commits.webkit.org/258742@main
